### PR TITLE
Selftype - fallback to type(x) for classmethods

### LINKED ISSUE #3630: Generic Self Type classmethods behave incorrectly on instances
Repro:

```python
from typing import TypeVar, Tuple, Type

T = TypeVar('T', bound='Friend')

class Friend:
    other = None  # type: Friend

    @classmethod
    def make(cls: Type[T]) -> T:
        return cls()

class SuperFriend(Friend):
    pass

reveal_type(SuperFriend.make())  # Revealed type is 'test_generic_self.SuperFriend*'
reveal_type(SuperFriend().make())  # Revealed type is '<nothing>'
```

Both invocations should return `test_generic_self.SuperFriend*`.

### PR DESCRIPTION
**Pull request by elazarg**
Fix #3630

### CODE REVIEW

File: mypy/checkmember.py [hunk 1]
```diff
@@ -472,7 +472,7 @@ class B(A): pass
         tvars = [TypeVarDef(n, i + 1, [], builtin_type('builtins.object'), tv.variance)
                  for (i, n), tv in zip(enumerate(info.type_vars), info.defn.type_vars)]
         if is_classmethod:
-            t = bind_self(t, original_type)
+            t = bind_self(t, original_type, is_classmethod=True)
```
Comment (elazarg) [inline mypy/checkmember.py]: This path is not tested

File: mypy/checkmember.py [hunk 2]
```diff
@@ -642,8 +642,13 @@ class B(A): pass
             # XXX value restriction as union?
             original_type = erase_to_bound(self_param_type)
 
-        typearg = infer_type_arguments([x.id for x in func.variables],
-                                       self_param_type, original_type)[0]
+        ids = [x.id for x in func.variables]
+        typearg = infer_type_arguments(ids, self_param_type, original_type)[0]
+        if (is_classmethod and isinstance(typearg, UninhabitedType)
```
Comment (ilevkivskyi) [inline mypy/checkmember.py]: Do you really need this additional test for ``UninhabitedType``? It looks a bit suspicious.

File: mypy/checkmember.py [hunk 3]
```diff
@@ -642,8 +642,13 @@ class B(A): pass
...
+                and isinstance(original_type, (Instance, TypeVarType, TupleType))):
+            # In case we call a classmethod through an instance x, fallback to type(x)
+            # TODO: handle Union
+            typearg = infer_type_arguments(ids, self_param_type, TypeType(original_type))[0]
```
Comment (ilevkivskyi) [inline mypy/checkmember.py]: Does actually ``Type[]`` accepts ``Tuple`` as an argument? I thought it works with unions, but not with tuple types. Please correct me if I am wrong. (Or is this only to cover self types in named tuples?)

File: test-data/unit/check-selftype.test [hunk 4]
```diff
@@ -219,13 +218,26 @@ class C:
...
+    reveal_type(arg.new)  # E: Revealed type is 'def () -> Q`-1'
+    reveal_type(arg.new())  # E: Revealed type is 'Q`-1'
+    return arg.copy()
 
-def make(cls: Type[T]) -> T:
-    reveal_type(cls.new)  # E: Revealed type is 'def () -> T`-1'
+def make(cls: Type[Q]) -> Q:
+    reveal_type(cls.new)  # E: Revealed type is 'def () -> Q`-1'
+    reveal_type(cls().new)  # E: Revealed type is 'def () -> Q`-1'
+    reveal_type(cls().new())  # E: Revealed type is 'Q`-1'
```
Comment (ilevkivskyi) [inline test-data/unit/check-selftype.test]: I would add more tests like in the original issue, i.e. with classes and ``@classmethod``.

File: mypy/checkmember.py (hunk 2 above)
Comment (elazarg) [inline mypy/checkmember.py]: Even if it is a classmethod, if the inference is successful with the original type, then it's fine; this is the normal case, calling the method on a class. We only "delegate" to the type if we are not successful.

File: mypy/checkmember.py (hunk 3 above)
Comment (elazarg) [inline mypy/checkmember.py]: I basically mirrored the restriction on the type of the parameters :) but yes, I think it is needed for namedtuples.

File: test-data/unit/check-selftype.test (hunk 4 above)
Comment (elazarg) [inline test-data/unit/check-selftype.test]: I'm not sure I understand. Can you elaborate? I thought the test for `D` above (line 226) is exactly the same as in the original issue.

Comment (ilevkivskyi) [inline test-data/unit/check-selftype.test]: Yes, sorry, GitHub just cut ``@classmethod`` for me.

File: mypy/checkmember.py (hunk 2 above)
Comment (ilevkivskyi) [inline mypy/checkmember.py]: OK, this seems reasonable. If there will be no objections, then I am going to merge this.

**Pull request merged**